pkg/infoschema/infoschema_v2.go

```diff
@@ -0,0 +1,431 @@
+
+func (is *infoschemaV2) AllSchemas() (schemas []*model.DBInfo) {
+	is.schemaMap.Scan(func(item schemaItem) bool {
+		// TODO: version?
+		schemas = append(schemas, item.dbInfo)
+		return true
+	})
+	return
+}
+
+func (is *infoschemaV2) SchemaMetaVersion() int64 {
+	return is.schemaVersion
+}
+
+func (is *infoschemaV2) SchemaExists(schema model.CIStr) bool {
+	var ok bool
+	// TODO: support different version
+	is.schemaMap.Scan(func(item schemaItem) bool {
+		if item.dbInfo.Name.L == schema.L {
+			ok = true
+			return false
+		}
+		return true
+	})
+	return ok
+}
+
+func (is *infoschemaV2) FindTableByPartitionID(partitionID int64) (table.Table, *model.DBInfo, *model.PartitionDefinition) {
+	panic("TODO")
+}
+
+func (is *infoschemaV2) TableExists(schema, table model.CIStr) bool {
+	_, err := is.TableByName(schema, table)
+	return err == nil
+}
+
+func (is *infoschemaV2) SchemaByID(id int64) (*model.DBInfo, bool) {
+	var ok bool
+	var dbInfo *model.DBInfo
+	is.schemaMap.Scan(func(item schemaItem) bool {
+		if item.dbInfo.ID == id {
+			ok = true
+			dbInfo = item.dbInfo
+			return false
+		}
+		return true
+	})
+	return dbInfo, ok
+}
+
+func (is *infoschemaV2) SchemaTables(schema model.CIStr) (tables []table.Table) {
+	dbInfo, ok := is.SchemaByName(schema)
+	if !ok {
+		return
+	}
+	snapshot := is.r.Store().GetSnapshot(kv.NewVersion(is.ts))
+	// Using the KV timeout read feature to address the issue of potential DDL lease expiration when
+	// the meta region leader is slow.
+	snapshot.SetOption(kv.TiKVClientReadTimeout, uint64(3000)) // 3000ms.
+	m := meta.NewSnapshotMeta(snapshot)
+	tblInfos, err := m.ListSimpleTables(dbInfo.ID)
+	if err != nil {
+		if meta.ErrDBNotExists.Equal(err) {
+			return nil
+		}
+		panic(err)
+	}
+	tables = make([]table.Table, 0, len(tblInfos))
+	for _, tblInfo := range tblInfos {
+		tbl, ok := is.TableByID(tblInfo.ID)
+		if !ok {
+			// what happen?
+			continue
+		}
+		tables = append(tables, tbl)
+	}
+	return
+}
+
+func loadTableInfo(r autoid.Requirement, infoData *Data, tblID, dbID int64, ts uint64) (table.Table, error) {
+	// Try to avoid repeated concurrency loading.
+	res, err, _ := sf.Do(fmt.Sprintf("%d-%d", dbID, tblID), func() (ret any, err error) {
+		snapshot := r.Store().GetSnapshot(kv.NewVersion(ts))
+		// Using the KV timeout read feature to address the issue of potential DDL lease expiration when
+		// the meta region leader is slow.
+		snapshot.SetOption(kv.TiKVClientReadTimeout, uint64(3000)) // 3000ms.
+		m := meta.NewSnapshotMeta(snapshot)
+
+		ret, err = m.GetTable(dbID, tblID)
+		return
+	})
+
+	if err != nil {
+		// TODO load table panic!!!
+		panic(err)
+	}
+	tblInfo := res.(*model.TableInfo) // TODO: it could be missing!!!
+
+	ConvertCharsetCollateToLowerCaseIfNeed(tblInfo)
+	ConvertOldVersionUTF8ToUTF8MB4IfNeed(tblInfo)
+	allocs := autoid.NewAllocatorsFromTblInfo(r, dbID, tblInfo)
+	b := NewBuilder(r, nil, infoData) // TODO: handle cached table!!!
+	ret, err := b.tableFromMeta(allocs, tblInfo)
```

Comment:
odd to attach a util method to a `Builder` obj

Reply:
Well, cached table need some post operations ...